lib/cog/chat/hipchat/connector.ex

```diff
@@ -0,0 +1,387 @@
+            {{:ok, user}, state} ->
+              GenServer.cast(state.provider, {:chat_message, %Cog.Chat.Message{id: Cog.Events.Util.unique_id,
+                                                                               room: room, user: user, text: body, provider: @provider_name,
+                                                                               bot_name: "@#{state.mention_name}", edited: false}})
+              state
+            error ->
+              Logger.error("Failed to lookup sender of groupchat message: #{inspect error}")
+              state
+          end
+        error ->
+          Logger.error("Failed to lookup room for groupchat message: #{inspect error}")
+      end
+    else
+      state
+    end
+  end
+
+  defp handle_dm(sender, body, state) do
+    case lookup_user(state, jid: sender) do
+      {{:ok, nil}, state} ->
+        state
+      {{:ok, user}, state} ->
+        room = %Cog.Chat.Room{id: sender,
+                              is_dm: true,
+                              name: "direct",
+                              provider: @provider_name}
+        GenServer.cast(state.provider, {:chat_message, %Cog.Chat.Message{id: Cog.Events.Util.unique_id,
+                                                                         room: room, user: user, text: body, provider: @provider_name,
+                                                                         bot_name: "@#{state.mention_name}", edited: false}})
+        state
+    end
+  end
+
+  defp send_output(state, target, output) do
+    case lookup_user(state, try_both: target) do
+      {{:ok, nil}, state} ->
+        case lookup_room(state, name: target) do
+          {{:ok, nil}, state} ->
+            Logger.warn("Unknown message target '#{target}'. Message NOT sent.")
+            {:reply, :ok, state}
+          {{:ok, room}, state} ->
+            send_room_message(room, output, state)
+          error ->
+            Logger.error("Failed to lookup target '#{target}' as room. Message NOT sent: #{inspect error}")
+        end
+      {{:ok, user}, state} ->
+        send_user_message(user, output, state)
+      error ->
+        Logger.error("Failed to lookup target '#{target}' as user. Message NOT sent: #{inspect error}")
+    end
+  end
+
+  defp prepare_message(text) when is_binary(text) do
+    Poison.encode!(%{message_format: "html",
+                     color: "gray",
+                     notify: false,
+                     message: text})
+  end
+
+  defp send_room_message(room, message, state) do
+    body = prepare_message(message)
+    url = Enum.join([state.api_root, "room", room.secondary_id, "notification"], "/") <> "?token=#{state.api_token}"
+    response = HTTPotion.post(url, headers: ["Content-Type": "application/json",
+                                             "Accepts": "application/json",
+                                             "Authorization": "Bearer #{state.api_token}"],
+      body: body)
```

Comment:
Might also be nice to wrap up this request stuff into a helper function. Totally not necessary though.